src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java

```diff
@@ -152,6 +160,50 @@ static <T> Blob constructBlobAndMetadata(
     return new Blob(blobBytes, chunksMetadataList, new BlobStats());
   }
 
+  /**
+   * Safety check to verify whether the number of rows in the parquet footer matches the number of
+   * rows in metadata
+   */
+  static <T> void verifyRowCounts(
+      Flusher.SerializationResult serializationResult,
+      List<ChannelData<T>> channelsDataPerTable,
+      byte[] paddedChunkData,
+      int chunkLength) {
```

Comment:
Another idea would be to add the check in the Parquet flusher inside `serializeFromX` methods right before we return the serialization result. We can get the per block row counts from the `parquetWriter` (through a new method) directly since the writer has access to the footer. So we wouldn't need to read and create a new `BdecInputFile`.

Reply:
Thanks, this simplifies the code pretty significantly. I updated the PR.